lib/object-path-immutable-transaction.js

```diff
@@ -0,0 +1,98 @@
+'use strict';
+const _assign = require('lodash.assign');
+const objectPath = require('object-path');
+const _isArray = require('lodash.isarray');
+const _isPlainObject = require('lodash.isplainobject');
+
+function ensureCloned(path, source, dest) {
+	if (!_isArray(path)) {
+		path = ('' + path).split('.');
+	}
+
+	// ensure all ancestor path segments are cloned
+	for (let i = 0; i < path.length; i++) {
+		if (dest[path[i]] === source[path[i]]) {
+			if (_isArray(source[path[i]])) {
+				dest[path[i]] = source[path[i]].slice();
+			} else {
+				dest[path[i]] = _assign({}, source[path[i]]);
+			}
+		}
+
+		source = source[path[i]];
+		dest = dest[path[i]];
+	}
+}
+
+function createTransaction(source) {
+	if (!_isPlainObject(source)) {
+		throw new Error('Source must be a plain object. Got ' + source);
+	}
+
+	const result = _assign({}, source);
```

Comment:
> Doesn't ensureCloned clone each object in source every time it is called?

No. See [L14](https://github.com/RetailMeNotSandbox/redux-mount-store/pull/9/files#diff-90bf56f455e9ff0dc8043fef9a75c86aR14), which checks whether the value in the result object has already been cloned.

> Is a goal of this PR to minimize how many times the source object and its children are cloned?

Not precisely, no. The goal is to fix the O(N^2) complexity encountered when performing repeated immutable operations using object-path-immutable. This code does so by not cloning any object or array in the result more than once, with the upshot that performing N operations is O(N), or O(N+M) in the worst case that the N operations cause all M nodes in the source object to be copied.

As you point out in your example, it's still possible that some clone will later turn out to be unnecessary. Determining that a given operation in a transaction is rendered moot by a later transaction seems like it might be hard to do correctly and efficiently.

> The referenced issue #7 discusses being lazy as part of the solution, but this code does not seem to be lazy.

Correct, it is not lazy. "Be lazy" in #7 refers to deferring application of mounts and unmounts until someone wants to read state affected by them, thus allowing us to batch.

When that batch is processed, we still need to ensure that we clone the ancestors of any mounted or unmounted paths. This code provides a transaction API that does so efficiently.

> Would something like this benefit from lodash's shortcut fusion that gets applied when evaluating chained methods?

My understanding is that shortcut fusion only works for [certain pipelines of operations that produce arrays](http://gajus.com/blog/4/harder-better-faster-stronger-lo-dash-v3#performance-improvements).